### PR TITLE
3.data

### PR DESCRIPTION
**Pull request by magdalena-u**
chapter: Data Structures: Objects and Arrays:
The sum of a range
Reversing an array
A list
Deep comparison

### CODE REVIEW

File: compare.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const deepEqual = (value1, value2) => {
+    if (typeof (value1) === typeof (value2)) {
+        if (typeof (value1 === "object")) {
+            (JSON.stringify(Object.keys(value1)) === JSON.stringify((Object.keys(value2))) ? true : false)
```
Comment (grzegorz-bielski) [inline compare.js]: You don't need all this junk (ifs) if you are using `JSON.stringifiy`... but don't use it, it's slow and will take into account ordering of the properties. TIP: you'll need some recursion and proper base cases

File: list.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const arrayToList = (arr) => {
+    let list;
+    for (let i = 0; i < arr.length; i++) {
+        list = {
+            value: arr[i],
+            rest: list
+        }
+    }
+    return list
```
Comment (grzegorz-bielski) [inline list.js]: The solution is fine, but It's a little bit hard to follow this list reassignment. This exercise is just perfectly suited for recursion.
```javascript
const arrayToList = ([value, ...values]) => 
	({ value, rest: values.length === 0 ? void 0 : arrayToList(values) })
```

File: list.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+const listToArray = (list) => {
+    let arr = []
+    for (let num = list; num; num = num.rest) {
+        arr.push(num.value)
```
Comment (grzegorz-bielski) [inline list.js]: Again, It's fine. (though I don't really like these reassignments in loops, it's hard to follow) 
Here is how this could be implemented using recursion in a pure way, just for the reference.
```javascript
const listToArray = ({ value, rest }, acc = []) => {
	const cumulated  = [...acc, value]
	return rest.length === 0 ? cumulated : listToArray(rest, cumulated)
}
```

File: reverse.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+const reverseArray = arr => (result = arr.map(i => i).reverse());
+
+const reverseArrayInPlace = arr => arr.reverse();
```
Comment (grzegorz-bielski) [inline reverse.js]: As far as I know, the exercise forbids using `Array.prototype.reverse`: https://eloquentjavascript.net/04_data.html#p_+SVCOC/qcN Implement it yourself :P

File: sum.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const range = (start = 0, end) => {
+    return [...Array(end - start).keys()].map(i => i + start);
```
Comment (grzegorz-bielski) [inline sum.js]: Oh, I forgot that `Array` have `keys` method, nice. (Also, you can omit `return` and braces)

File: sum.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const range = (start = 0, end) => {
```
Comment (grzegorz-bielski) [inline sum.js]: Default parameters only make sense if they are last

File: sum.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+    return [...Array(end - start).keys()].map(i => i + start);
+}
+
+const sum = (range) => {
```
Comment (grzegorz-bielski) [inline sum.js]: You can omit `()` braces if the arrow function is taking only one argument

**Review by michal-karbasz — APPROVED** [review]
